sunpy/map/maphelper.py

```diff
@@ -0,0 +1,246 @@
+    >>> from sunpy import map
+    >>> from sunpy.coordinates import frames
+    >>> from astropy.coordinates import SkyCoord
+    >>> from astropy import units as u
+    >>> import numpy as np
+
+    >>> data = np.random.rand(1024, 1024)
+    >>> my_coord = SkyCoord(0*u.arcsec, 0*u.arcsec, obstime="2017-08-01", observer = 'earth', frame=frames.Helioprojective)
+    >>> my_header = map.make_fits_wcs(data, my_coord)
+    >>> my_map = map.Map(data, my_header)
+    """
+
+    if not isinstance(coordinate, SkyCoord):
+        raise ValueError("coordinate needs to be a SkyCoord instance.")
+
+    if coordinate.obstime is None:
+        raise ValueError("The coordinate needs an observation time, `obstime`.")
+
+    coordinate = coordinate.transform_to(frames.Helioprojective)
+
+    meta_wcs = _get_wcs_meta(coordinate)
+
+    meta_observer = _get_observer_meta(coordinate)
+
+    meta_instrument = _get_instrument_meta(**kwargs)
+
+    meta_wcs.update(meta_observer)
+    meta_wcs.update(meta_instrument)
+
+    if reference_pixel is None:
+        reference_pixel = u.Quantity([(data.shape[1] + 1)/2.*u.pixel, (data.shape[0] + 1)/2.*u.pixel])
+    if scale is None:
+        scale = u.Quantity([1.0*u.arcsec, 1.0*u.arcsec])
+
+    meta_wcs['crval1'], meta_wcs['crval2'] = coordinate.Tx.to_value(meta_wcs['cunit1']), coordinate.Ty.to_value(meta_wcs['cunit2'])
+
+    meta_wcs['crpix1'], meta_wcs['crpix2'] = reference_pixel[0].to_value(u.pixel), reference_pixel[1].to_value(u.pixel)
+
+    meta_wcs['cdelt1'], meta_wcs['cdelt2'] = scale[0]*meta_wcs['cunit1']/u.pixel, scale[1]*meta_wcs['cunit2']/u.pixel
+
+    meta_dict = MetaDict(meta_wcs)
+
+    for key in kwargs:
+        if key in map_meta_keywords:
+            meta_dict[key] = kwargs[key]
+
+    return meta_dict
+
+
+def _get_wcs_meta(coordinate):
+    """
+    Function to get WCS meta from the SkyCoord using
+    `astropy.wcs.utils.celestial_frame_to_wcs`
+
+    Parameters
+    ----------
+    coordinate : ~`astropy.coordinates.SkyCoord`
+
+    Returns
+    -------
+    `dict`
+        Containing the WCS meta information
+            * ctype1, ctype2
+            * cunit1, cunit2
+            * date_obs
+    """
+
+    coord_meta = {}
+
+    skycoord_wcs = astropy.wcs.utils.celestial_frame_to_wcs(coordinate.frame)
+
+    cunit1, cunit2 = skycoord_wcs.wcs.cunit
+    coord_meta = dict(skycoord_wcs.to_header())
+    coord_meta['cunit1'], coord_meta['cunit2'] = cunit1, cunit2
+
+    return coord_meta
+
+
+def _get_observer_meta(coordinate):
+    """
+    Function to get observer meta from the SkyCoord using
+    `astropy.wcs.utils.celestial_frame_to_wcs`
+
+    Parameters
+    ----------
+    coordinate : ~`astropy.coordinates.SkyCoord`
+
+    Returns
+    -------
+    `dict`
+        Containing the WCS meta information
+            * hgln_obs, hglt_obs
+            * dsun_obs
+            * rsun_obs
+    """
+    coord_meta = {}
+
+    skycoord_wcs = astropy.wcs.utils.celestial_frame_to_wcs(coordinate.frame)
```

Comment:
Do we want this function to also do this? My understanding after our chat on Wed was this one would *only* do the observer stuff?

Reply:
yeah so at the moment it just does the observer stuff right? It just pulls it from the heliographic_observer info in the frame - should we pull it rather from the skycoord observer keyword rather than the frame?

Reply:
Yeah I would suggest you use the coordinate directly.

Reply:
ping @Cadair